### PR TITLE
Fix EscapeAnalysis verification assert at unreachable blocks

### PR DESCRIPTION
**Pull request by atrick**
If EscapeAnalysis verification runs on unreachable code, it asserts
with "Missing escape connection graph mapping" because the connection
graph builder only runs on reachable blocks.

Add a ReachableBlocks utility and use it during verification.

Fixes <rdar://problem/60373501> EscapeAnalysis crashes with CFG with
unreachable blocks

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci smoke test linux

File: include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h [hunk 1]
```diff
@@ -33,6 +33,24 @@ class BasicBlockCloner;
 class SILLoop;
 class SILLoopInfo;
 
+/// Compute the set of reachable blocks.
+class ReachableBlocks {
+  llvm::SmallPtrSet<SILBasicBlock *, 32> visited;
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h]: Doesn't need llvm::.

File: include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h [hunk 2]
```diff
@@ -33,6 +33,24 @@ class BasicBlockCloner;
...
+
+public:
+  /// Invoke \p visitor for each reachable block in \p f in worklist order (at
+  /// least one predecessor has been visited--defs are always visited before
+  /// uses except for phi-type block args). The \p visitor takes a block
+  /// argument, which is already marked visited, and must return true to
+  /// continue visiting blocks.
+  ///
+  /// Returns true if all reachable blocks were visited.
+  bool visit(SILFunction *f, std::function<bool(SILBasicBlock *)> visitor);
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h]: Shouldn't this be an function_ref? You aren't going to ever store/escape the std::function. I guess it is better to be conservative though.

File: lib/SILOptimizer/Utils/BasicBlockOptUtils.cpp [hunk 3]
```diff
@@ -17,6 +17,30 @@
...
+/// Invoke \p visitor for each reachable block in \p f in worklist order (at
+/// least one predecessor has been visited).
+bool ReachableBlocks::visit(SILFunction *f,
+                            std::function<bool(SILBasicBlock *)> visitor) {
+  assert(visited.empty() && "blocks already visited");
+
+  // Walk over the CFG, starting at the entry block, until all reachable blocks
+  // are visited.
+  SILBasicBlock *entryBB = f->getEntryBlock();
+  llvm::SmallVector<SILBasicBlock *, 8> worklist = {entryBB};
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/BasicBlockOptUtils.cpp]: llvm:: is not needed.

**Review by gottesmm — APPROVED** [review]
A few small nits. LGTM otherwise.

**Comment by atrick** [comment]
@swift-ci smoke test

File: include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h (hunk 2 above)
Comment (atrick) [reply include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h]: It can be `function_ref`, I didn't think carefully enough about it.

**Pull request merged**